From 7acd744fb5ce9d69a825fab7a9772b86b7b362d5 Mon Sep 17 00:00:00 2001
From: Jiangli Yuan <jlyuan@motorola.com>
Date: Mon, 25 Apr 2016 16:09:09 -0700
Subject: [PATCH 867/959] IKSWM-33312: MMC: trigger kernel panic if too many
 CMDQ timeout

 Trigger kernel panic if there are 10 cmdq timeout within 5 mins.

Change-Id: Ic17013ffcb59763836b40e2a1bd780a043af00bf
Signed-off-by: Jiangli Yuan <jlyuan@motorola.com>
Reviewed-on: http://gerrit.mot.com/848541
SLTApproved: Slta Waiver <sltawvr@motorola.com>
SME-Granted: SME Approvals Granted
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Zhi-Ming Yuan <a14194@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/mmc/card/queue.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/drivers/mmc/card/queue.c b/drivers/mmc/card/queue.c
index 9ff67261f3b..abc92fefa2e 100644
--- a/drivers/mmc/card/queue.c
+++ b/drivers/mmc/card/queue.c
@@ -659,10 +659,34 @@ static void mmc_cmdq_error_work(struct work_struct *work)
 enum blk_eh_timer_return mmc_cmdq_rq_timed_out(struct request *req)
 {
 	struct mmc_queue *mq = req->q->queuedata;
+	static ktime_t last_timeout;
+	static int mmc_cmdq_req_timeout_count;
+	ktime_t now;
+	s64 delta;
+
 
 	pr_err("%s: request with tag: %d flags: 0x%x timed out\n",
 	       mmc_hostname(mq->card->host), req->tag, req->cmd_flags);
 
+	if (!mmc_cmdq_req_timeout_count) {
+		last_timeout = ktime_get();
+		mmc_cmdq_req_timeout_count = 1;
+	} else {
+		now = ktime_get();
+		delta = ktime_to_ms(ktime_sub(now, last_timeout));
+
+		/* count only if two sequenet requests are
+		   timeout within 5 mins */
+		if (delta < 300000)
+			mmc_cmdq_req_timeout_count++;
+		else
+			mmc_cmdq_req_timeout_count = 0;
+		last_timeout = now;
+	}
+
+	if (mmc_cmdq_req_timeout_count >= 10)
+		BUG();
+
 	return mq->cmdq_req_timed_out(req);
 }
 
-- 
2.11.0

